### PR TITLE
Support multi node join hints

### PR DESCRIPTION
**Pull request by systay**

### CODE REVIEW

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/ast/Hint.scala [hunk 1]
```diff
@@ -46,18 +47,31 @@ sealed trait UsingHint extends Hint
...
+  def identifiers = NonEmptyList(identifier)
   def semanticCheck = identifier.ensureDefined chain identifier.expectType(CTNode.covariant)
 }
 
-case class UsingJoinHint(identifier: Identifier)(val position: InputPosition) extends UsingHint with NodeHint {
-  def semanticCheck = identifier.ensureDefined chain identifier.expectType(CTNode.covariant)
+object UsingJoinHint {
+  import NonEmptyList._
+
+  def apply(elts: Seq[Identifier] = Seq.empty)(pos: InputPosition): UsingJoinHint =
```
Comment (pontusmelke) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/ast/Hint.scala]: Mostly curious, why the default value here?

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/planner/logical/greedy/join.scala [hunk 2]
```diff
@@ -42,7 +42,7 @@ object join extends CandidateGenerator[GreedyPlanTable] {
         if (shared.nonEmpty && shared.forall(isApplicable)) {
 
           val hints = queryGraph.hints.collect {
-            case hint@UsingJoinHint(identifier) if shared contains IdName(identifier.name) => hint
+            case hint@UsingJoinHint(identifiers) if identifiers.forall(identifier => shared contains IdName(identifier.name)) => hint
```
Comment (pontusmelke) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/planner/logical/greedy/join.scala]: Isn't this just `shared == identifiers.map(_.name).toSet`
hmm...maybe not much simpler

File: community/cypher/docs/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/UsingTest.scala [hunk 3]
```diff
@@ -68,10 +79,26 @@ class UsingTest extends DocumentingTestBase {
...
+      title = "Hinting a join on a single node",
       text = "To force the query planner to produce plans with joins in them, use +USING+ +JOIN+.",
       queryText = "match (andres {name:'Andres'})-->(x)<--(emil {name: 'Emil'}) using join on x return x",
       assertions = (p) => assert(p.toList === List(Map("x" -> node("Peter"))))
     )
   }
+
+  @Test def query_forcing_join_using_multiple_nodes() {
+    profileQuery(
+      title = "Hinting a join on multiple nodes",
```
Comment (pontusmelke) [inline community/cypher/docs/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/UsingTest.scala:91]: Changing the title also forces you to change the include in https://github.com/neo4j/neo4j/blob/2.3/community/cypher/docs/cypher-docs/src/docs/dev/ql/using/index.asciidoc, I think that's why build is failing.

Comment (boggle) [inline community/cypher/docs/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/UsingTest.scala:91]: Fixed.

**Comment by systay** [comment]
Last commit on this PR is already merged

**Pull request merged**